pkg/monitor/main.go

```diff
@@ -0,0 +1,172 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/daemon/defaults"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/byteorder"
+	"github.com/cilium/cilium/pkg/monitor/payload"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const targetName = "cilium-node-monitor"
+
+func receiveEvent(m *bpf.PerfEventSample, c int) {
+	pl := payload.Payload{Data: m.DataCopy(), CPU: c, Lost: 0, Type: payload.EventSample}
+	send(pl)
+}
+
+func lostEvent(m *bpf.PerfEventLost, c int) {
+	pl := payload.Payload{Data: []byte{}, CPU: c, Lost: m.Lost, Type: payload.RecordLost}
+	send(pl)
+}
+
+var (
+	listeners     = make(map[int]net.Conn)
+	mutex         sync.Mutex
+	monitorEvents *bpf.PerCpuEvents
+)
+
+// handleMonitorConn handles all the incoming monitors.
+func handleMonitorConn(server net.Listener) {
+	for {
+		conn, err := server.Accept()
+		if err != nil {
+			log.Print(err)
+			continue
+		}
+
+		mutex.Lock()
+		listeners[len(listeners)+1] = conn
+		log.Infof("New monitor connected. len(listeners) = %d", len(listeners))
+		mutex.Unlock()
+	}
+}
+
+func dumpStat() {
+	c := int64(monitorEvents.Cpus)
+	n := int64(monitorEvents.Npages)
+	p := int64(monitorEvents.Pagesize)
+	l, u := monitorEvents.Stats()
+	m := models.MonitorStatus{Cpus: c, Npages: n, Pagesize: p, Lost: int64(l), Unknown: int64(u)}
+
+	mp, err := json.Marshal(m)
+	if err != nil {
+		log.Errorf("json: %s", err)
+		return
+	}
+	fmt.Println(string(mp))
+}
+
+// runMonitor starts monitoring.
+func runMonitor(npages int) {
+	log.Info("Starting monitoring traffic")
+	c := bpf.DefaultPerfEventConfig()
+	c.NumPages = npages
+
+	me, err := bpf.NewPerCpuEvents(c)
+	if err != nil {
+		log.Errorf("Error while starting monitor: %s", err)
+		return
+	}
+	monitorEvents = me
+
+	// Dump stat
+	t := time.NewTicker(5 * time.Second)
+	go func(t *time.Ticker) {
+		for {
+			select {
+			case <-t.C:
+				if monitorEvents != nil {
+					dumpStat()
+				}
+			}
+		}
+	}(t)
+
+	// Main event loop
+	for {
+		todo, err := monitorEvents.Poll(5000)
+		if err != nil {
+			log.Errorf("Error Poll %s", err)
+			if err == syscall.EBADF {
+				break
+			}
+		}
+		if todo > 0 {
+			if err := monitorEvents.ReadAll(receiveEvent, lostEvent); err != nil {
+				log.Warningf("Error received while reading from perf buffer: %s", err)
+			}
+		}
+	}
+}
+
+func send(pl payload.Payload) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	if len(listeners) == 0 {
+		return
+	}
+
+	buf := pl.Encode()
+	meta := &payload.Meta{Size: uint32(len(buf))}
+	for k, client := range listeners {
```

Comment:
Just have to be careful that Remove(e) sets e.Next() to nil. Use another variable to hold the next element before removing:

    for e := listeners.Front(); e != nil; {
        next := e.Next()
        client := e.Value().(net.Conn)
        ...
        // To delete:
        listeners.Remove(e)
        ...
        e = next
    }